### ISSUE TITLE
Mypy handling initializing Pydantic models with dict values incorrectly

### ISSUE THREAD

**Issue by dpitch40** [python/mypy#17244]
**Bug Report**

Mypy is flagging code pasted directly from the Pydantic docs (involving initializing models with dictionary values for nested classes) as incorrectly typed.

**To Reproduce**

Save the following code (directly from the [current Pydantic docs](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions-with-str-discriminators) to a Python file.
 
```python
from typing import Literal, Union

from pydantic import BaseModel, Field, ValidationError


class Cat(BaseModel):
    pet_type: Literal['cat']
    meows: int


class Dog(BaseModel):
    pet_type: Literal['dog']
    barks: float


class Lizard(BaseModel):
    pet_type: Literal['reptile', 'lizard']
    scales: bool


class Model(BaseModel):
    pet: Union[Cat, Dog, Lizard] = Field(..., discriminator='pet_type')
    n: int


print(Model(pet={'pet_type': 'dog', 'barks': 3.14}, n=1))
```

Then run mypy on it.

**Expected Behavior**

Mypy should validate the above code without errors, as it is example code directly from Pydantic's docs.

**Actual Behavior**

```
test.py:26: error: Argument "pet" to "Model" has incompatible type "dict[str, object]"; expected "Cat | Dog | Lizard"  [arg-type]
Found 1 error in 1 file (checked 1 source file
```

**Your Environment**

- Mypy version used: 1.10.0
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): [mypy.ini](https://github.com/python/mypy/files/15312688/mypy.ini.txt)
- Python version used: 3.10.11
- Pydantic version used: 2.7.1

**Comment by dpitch40** [comment]
I can also reproduce it with [this code](https://docs.pydantic.dev/latest/concepts/models/#nested-models) from the Pydantic docs.

```python
from typing import List, Optional

from pydantic import BaseModel


class Foo(BaseModel):
    count: int
    size: Optional[float] = None


class Bar(BaseModel):
    apple: str = 'x'
    banana: str = 'y'


class Spam(BaseModel):
    foo: Foo
    bars: List[Bar]


m = Spam(foo={'count': 4}, bars=[{'apple': 'x1'}, {'apple': 'x2'}])
```

The Pydantic docs appear to treat code initializing nested classes using dictionaries containing the fields and values as correct (and the code does work for me if I run it), but Mypy is flagging such code as incorrect.

**Comment by hauntsaninja** [comment]
mypy isn't going to support this. I think pydantic has a plugin that might help with this, you could ask on their repo.
